mne/time_frequency/tests/test_spectrum.py

```diff
@@ -295,44 +432,135 @@ def _check_spectrum_equivalent(spect1, spect2, tmp_path):
+    # test mismatching number of channels
+    bad_n_chans = data[:-1] if kind == "raw" else data[:, :-1]
     with pytest.raises(ValueError, match=r"number of channels.*good data channels"):
-        EpochsSpectrumArray(data[:, :-1], info, freqs)
-    with pytest.raises(ValueError, match=r"last dimension.*same number of elements"):
-        EpochsSpectrumArray(data[..., :-1], info, freqs)
+        Klass(bad_n_chans, info, freqs)
+    # test mismatching number of frequencies
+    bad_n_freqs = (
+        data[..., :-1, :] if method == "welch" and not average else data[..., :-1]
+    )
+    with pytest.raises(ValueError, match=r"number of frequencies.*number of elements"):
+        Klass(bad_n_freqs, info, freqs, method=method, mt_weights=mt_weights)
     # test mismatching events shape
-    n_epo = data.shape[0] + 1  # +1 so they purposely don't match
-    events = np.vstack(
-        (np.arange(n_epo), np.zeros(n_epo, dtype=int), np.ones(n_epo, dtype=int))
-    ).T
-    with pytest.raises(ValueError, match=r"first dimension.*dimension of `events`"):
-        EpochsSpectrumArray(data, info, freqs, events)
+    if kind == "epochs":
+        n_epo = data.shape[0] + 1  # +1 so they purposely don't match
+        events = np.vstack(
+            (np.arange(n_epo), np.zeros(n_epo, dtype=int), np.ones(n_epo, dtype=int))
+        ).T
+        with pytest.raises(ValueError, match=r"first dimension.*dimension of `events`"):
+            Klass(data, info, freqs, events)
+    # test unspecified method for unaggregated spectra (i.e. with segments or tapers)
+    if (
+        method == "welch"
+        and not average
+        or method == "multitaper"
+        and output == "complex"
+    ):
+        with pytest.raises(
+            ValueError, match="Invalid value for the 'method' parameter"
+        ):
+            Klass(data, info, freqs, method="unknown", mt_weights=mt_weights)
+    # test unspecified/mismatched multitaper weights
+    if method == "multitaper" and output == "complex":
+        with pytest.raises(
+            ValueError, match=r"Expected size of `mt_weights` to be.*, got"
+        ):
+            Klass(data, info, freqs, method=method, mt_weights=None)
+        with pytest.raises(
+            ValueError, match=r"Expected size of `mt_weights` to be.*, got"
+        ):
+            Klass(data, info, freqs, method=method, mt_weights=mt_weights[:, :-1])
 
 
 @pytest.mark.parametrize("kind", ("raw", "epochs"))
-def test_spectrum_array(kind, tmp_path, request):
+@pytest.mark.parametrize(
+    "method, output, average",
+    [
+        ("welch", "power", "mean"),  # test with precomputed spectrum
+        ("welch", "power", False),
+        ("welch", "complex", False),
+        ("welch", "complex", "mean"),
+        ("multitaper", "complex", None),
+    ],
+)
+def test_spectrum_array(kind, method, output, average, tmp_path, request):
     """Test EpochsSpectrumArray and SpectrumArray constructors."""
```

Comment:
For testing the `Array` constructors thoroughly, I needed complex & unaggregated (i.e. with tapers/segments) data beyond the existing `raw/epochs_spectrum` fixtures.

Is using the regular `raw/epochs` fixture and calling `compute_psd()` with the appropriate args okay, or would a `raw/epochs_spectrum_complex` ficture be preferred?

Reply:
good enough I think. fixtures are really meant for objects that will be re-used across more than one test, so since they're only used here (right?) you wouldn't save any compute time by making these fixtures.

Reply:
Not entirely:
- `("welch", "power", False), ("welch", "complex", False), ("welch", "complex", "mean"), ("multitaper", "complex", None)` are also used in `test_plot_spectrum`
- `("welch", "power", False), ("multitaper", "complex", None)` are also used in `test_spectrum_array_errors`

But actually thinking about it `test_spectrum_array_errors` doesn't need any genuine data so I could try to avoid excess run time there if you think it's worth.

Reply:
yeah... I think rather than creating fixtures we should test fewer cases. With the new parametrizations, this test went from being run twice to being run ten times, and `test_plot_spectrum` went from 4 times to 20 times. I think we want to be a bit strategic about which cases are really necessary for full coverage. E.g. I'd guess we don't need to test all complex cases with *both* raw and epochs objects, so maybe the parametrization of raw-vs-epochs gets rolled into the other one, and we do something like:

```
    "kind, method, output, average", 
    [
        ("epochs", "welch", "power", "mean"),
        ("raw", "welch", "power", "mean"),  # maybe even skip this one?
        ("raw", "welch", "power", False),
        ("raw", "welch", "complex", False),
        ("raw", "welch", "complex", "mean"),
        ("raw", "multitaper", "complex", None),
    ],
```
(so, 6 runs instead of 10). You can use your judgment about which ones are likely sufficient for full coverage, then check if coverage drops.

Reply:
Okay, I'll look into maintaining this coverage with fewest no. of tests.

Reply:
Okay, I'll look into maintaining this coverage with fewest no. of tests.

Reply:
Have trimmed down a lot of these for minimal re-runs whilst maintaining coverage (e.g. now only 3 runs each for `test_spectrum_array` and `test_plot_spectrum`).

I have slightly increased the size of `test_spectrum_array_errors` as a single test, however it now only needs to be run once (3 runs before) and is only using random data, so no more computing spectra for these non-fixture settings which will also help speed (a little).